Patterns/LayeredProject.md

```diff
@@ -0,0 +1,100 @@
+
+## Inter-project data exchange and isolation
+
+When dividing a solution into several distinct projects it is important to maintain a level of isolation between these projects. Keeping these projects reasonably isolated helps to control the impact of changes; a change is one project should not force a lot of changes in other projects. Isolation will prevent developers from having to notify users of the API of your solution that some outbound data changes due to modification in a database column. Keep in mind to not overdo this, implementation details will always tend to spread a little bit to other parts of the solution.
+
+### Isolate models
+
+Models, or any object containing data, are usually passed around between various parts of the application. Each project should use its own type of object for its own use.
+
+- An API always has its own Request- and ResponseModels. These models describe the exact shape of requests it expects and responses it generates.
+- The Core projects contains data transfer objects (DTOs) which facilitate communication between parts of the application.
+- An Gateway always has its own models. These models describe what data must be send to the external system and what data comes out of the external system. The gateway exposes an interface for use by the solution which accepts and returns DTOs.
+- The persistence layer uses its own entities to store data in a database or other storage, and exposes an interface which accepts and returns DTOs.
+- Business logic uses domain (or business) objects for its logic, or use DTOs directly.
+
+A payment request is made to the projects API controller, receiving an `AddPaymentRequestModel`. The API validates whether the request is valid (using data annotations on the model, or FluentValidation for that model). If the request is valid, the API controller transform the model to `AddPaymentDto` and passes it to `IPaymentService`. The implementer of this interface can opt to validate `AddPaymentDto`, but can also assume its validity. The service talks to `IPaymentGateway` which accepts the `AddPaymentDto`, converts it to `AdyenCreatePaymentModel` and sends it to the payment provider. The payment provider returns `PaymentModel`, which is converted to `PaymentDetailsDto` which is returned by the `IPaymentGateway`. The payment service uses that dto to talk to the `IPaymentRepository` to save the payment. That repository converts the `PaymentDetailsDto` to `PaymentEntity` which is persisted in the database.
+
+Each part has their own models for their own logic, isolating each part. A change in the shape of `AddPaymentRequestModel` will not force other services and interfaces to change. This makes modifying the application more easy. This way of working involves a lot of mapping between objects. A setup where these mappings are automatically generated is advisable and the use of things like AutoMapper, ExpressionMapper or GeneratedMapper is encouraged.
+
```

Comment:
In the example above, several boundaries have been drawn. Over these lines the Dependency Inversion principle is applies, isolating components. This helps improve the Single Responsibility Principle and avoids a tightly coupled system. It does require more specific models - one pair for each component - and the necessary mappers, for which you are advised to use a tool like [AutoMapper](https://automapper.org/). The biggest advantage of these boundaries are that you isolate components, and make your system loosely coupled. If for example, Adyen will be phased out in favor of another payment provider, you will only have to write a new class that implements IPaymentGateway and replace it for the Adyen-implementation.

Reply:
Is this text an addition to this text, or where do you want to put it?

Reply:
Addition, modification, as you see fit ;-)